File: app/src/main/java/jp/co/geisha/itunestracker/service/viewModel/ArtistListViewModel.java
package jp.co.geisha.itunestracker.service.viewModel;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.os.Handler;
import android.text.TextUtils;
import android.util.Log;

import jp.co.geisha.itunestracker.service.model.ArtistList;
import jp.co.geisha.itunestracker.service.repository.ArtistRepository;
import timber.log.Timber;

import static jp.co.geisha.itunestracker.service.util.CalcUtils.getRand;


public class ArtistListViewModel extends AndroidViewModel {

    private static final int DELAY_MINUTES = 60000;
    private static final int MAX_REQUEST_PER_MINUTE = 20;
    private int LIMIT = 100;
    private static final int ZERO = 0;
    private static final String MUSIC_VIDEO = "musicVideo";
    private LiveData<ArtistList> artistListObservable;
    private ArtistRepository repo;
    private int count;

    //TODO:決め打ちでなく、rankingからとってきてTrendとしてもいいかも
    private String[] defaultArtists =
            {
                    "alicia keys",
                    "lady gaga",
                    "michael jackson",
                    "beatles",
                    "stevie wonder",
                    "eric clapton",
                    "beyonce",
                    "james brown",
                    "sting",
                    "oasis",
                    "2pac",
                    "Nas",
                    "bob marley",
                    "billy joel",
                    "elton john",
                    "bruno mars",
                    "joe",
                    "justin timberlake",
                    "TLC",
                    "SWV",
                    "blackstreet",
                    "jackson5",
                    "ed sheeran",
                    "boyz 2 men",
                    "india arie",
                    "talor swift",
                    "norah jones",
                    "frank sinatra",
                    "marvin gaye",
                    "mariah carey",
                    "diana ross",
                    "jamiroquai",
                    "john lenon",
                    "tuxedo"
            };

    public ArtistListViewModel(Application application) {
        super(application);

        setUpRequestScheduler();

        repo = ArtistRepository.getInstance();

        artistListObservable = repo.getArtistList(defaultArtists[getRand(defaultArtists.length)], MUSIC_VIDEO, LIMIT);
    }

    //LiveDataのゲッター
    public LiveData<ArtistList> getArtistListObservable() {

        return artistListObservable;
    }

    public void reloadArtists(CharSequence text) {
        if (count <= MAX_REQUEST_PER_MINUTE) {
            if (TextUtils.isEmpty(text)) {
                artistListObservable = repo.getArtistList(defaultArtists[getRand(defaultArtists.length)], MUSIC_VIDEO, LIMIT);
            } else {
                artistListObservable = repo.getArtistList(text.toString(), "musicVideo", LIMIT);
                count++;
            }
        } else {
            Timber.d("%s制限超過", count);
        }
    }

    private void setUpRequestScheduler() {

        final Handler handler = new Handler();

        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                // UIスレッド
                if (count > MAX_REQUEST_PER_MINUTE) {
                    Timber.d("%sだから危険", count);
                } else Timber.d("%sだから安全", count);
                count = ZERO;
                handler.postDelayed(this, DELAY_MINUTES);
            }
        }, DELAY_MINUTES);
    }

}

File: app/src/main/java/jp/co/geisha/itunestracker/service/viewModel/ArtistListViewModel.kt
package jp.co.geisha.itunestracker.service.viewModel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import android.os.Handler
import android.text.TextUtils
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import jp.co.geisha.itunestracker.service.*
import jp.co.geisha.itunestracker.service.model.ArtistList
import jp.co.geisha.itunestracker.service.repository.ArtistRepository
import timber.log.Timber
import jp.co.geisha.itunestracker.service.ConstArrays.DEFAULT_ARTIST_LIST
import jp.co.geisha.itunestracker.service.util.CalcUtils.getRand
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class ArtistListViewModel(application: Application) : AndroidViewModel(application) {

    sealed class Status {
        object Success : Status()
        object Loading : Status()
        data class Error(val message: String) : Status()
    }

    val videoListLiveData: MutableLiveData<ArtistList> = MutableLiveData()
    val state: MutableLiveData<Status> = MutableLiveData()
    private val repo: ArtistRepository = ArtistRepository.instance
    private val handler = Handler()
    private var count: Int = 0

    init {
        setUpRequestScheduler()
    }

    fun loadArtists(query: CharSequence) {
        viewModelScope.launch {
            state.value = Status.Loading
            if (count <= MAX_REQUEST_PER_MINUTE) {
                if (TextUtils.isEmpty(query)) {
                    getVideoListWithQuery(DEFAULT_ARTIST_LIST[getRand(DEFAULT_ARTIST_LIST.size)])
                } else {
                    getVideoListWithQuery(query.toString())
                }
            } else {
                state.postValue(Status.Error("loadArtists :api call failed Possible limit exceeded , Now count == $count"))
            }
        }
    }

    private suspend fun getVideoListWithQuery(query: String) = withContext(Dispatchers.Default) {
        val result = repo.getMusicVideoList(query, MUSIC_VIDEO, LIMIT)
        if (result.isSuccessful) {
            state.postValue(Status.Success)
            videoListLiveData.postValue(result.body())
            count++
        } else {
            state.postValue(Status.Error("getVideoListWithQuery : api call failed"))
        }
    }

    val scheduler = object : Runnable {
        override fun run() {
            if (count > MAX_REQUEST_PER_MINUTE) {
                Timber.d("%s danger count", count)
            } else
                Timber.d("%s count safely", count)
            count = ZERO
            handler.postDelayed(this, DELAY_MINUTES.toLong())
        }
    }

    private fun setUpRequestScheduler() {
        viewModelScope.launch {
            handler.postDelayed(scheduler, DELAY_MINUTES.toLong())
        }
    }

    override fun onCleared() {
        super.onCleared()
        handler.removeCallbacks {
            scheduler
        }
    }
}
